### PR TITLE
Enable running Graylog REST API on different context path

### PR DESCRIPTION
**Pull request by joschi**
## Description

Refs #2603

## Motivation and Context

Users might (rightfully) want to run the Graylog REST API and the web interface on the same network listener with the web interface being served on the root context path (`/`) and the Graylog REST API being served from a path below that (`/api`).

This PR merges `RestApiService` and `WebInterfaceService` into a single service which handles setting up the required handlers accordingly.

## How Has This Been Tested?

Painfully.

Things that might still be broken and might only be working on my machine™:
* Swagger UI in general
* The generic 404-handler so that the web interface can have an arbitrary entry-point in the web browser
* Generated links in the Graylog REST API

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly
  - Well, the upgrade notes at least.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: graylog2-server/src/main/java/org/graylog2/shared/rest/resources/system/RemoteDeflectorResource.java [hunk 1]
```diff
@@ -20,6 +20,6 @@
 import retrofit2.http.POST;
 
 public interface RemoteDeflectorResource {
-    @POST("/system/deflector/cycle")
+    @POST("system/deflector/cycle")
```
Comment (kroepke) [inline graylog2-server/src/main/java/org/graylog2/shared/rest/resources/system/RemoteDeflectorResource.java]: Are these changes strictly necessary, or rather can we catch this in a central place?
I'm worried that plugins that implement a ProxiedResource will forget to make the path relative and thus break at runtime.

Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/shared/rest/resources/system/RemoteDeflectorResource.java]: Yes, these are necessary.

Otherwise the custom context path would be overridden (because the paths in these Retrofit resource classes are absolute and not relative as they should be). I haven't looked into it too deeply but I haven't found a suitable mechanism in [Retrofit's documentation](https://square.github.io/retrofit/) yet.

Also see [Retrofit 2 — Url Handling, Resolution and Parsing](https://futurestud.io/blog/retrofit-2-url-handling-resolution-and-parsing).

Comment (kroepke) [inline graylog2-server/src/main/java/org/graylog2/shared/rest/resources/system/RemoteDeflectorResource.java]: Ok bummer. I was asking because at least the audit log plugin needs the corresponding change in this case.

Comment (kroepke) [inline graylog2-server/src/main/java/org/graylog2/shared/rest/resources/system/RemoteDeflectorResource.java]: Could we potentially add a runtime check for the affected resources to log error messages?
Not sure where in the code this would have to go, the closest I could find is https://github.com/Graylog2/graylog2-server/blob/e638987fbf0e58e2211302a558a003fbfc18dff2/graylog2-server/src/main/java/org/graylog2/rest/RemoteInterfaceProvider.java#L41-L63

File: graylog2-server/src/main/resources/swagger/index.html.template [hunk 2]
```diff
@@ -1,98 +1,98 @@
...
+    var apiTarget = window.location.protocol + "//" + window.location.hostname + ":" + port + "${apiPrefix}";
+    $(function () {
+      window.swaggerUi = new SwaggerUi({
+      url: apiTarget + "api-docs",
+      dom_id: "swagger-ui-container",
+      supportedSubmitMethods: ['get', 'post', 'put', 'delete'],
+      onComplete: function(swaggerApi, swaggerUi){
+
+        // We are taking the base path of the system API here. This will break if you should ever rename it lol.
+        if(swaggerApi.System && !apiTarget.startsWith(swaggerApi.System.basePath))   {
```
Comment (kroepke) [inline graylog2-server/src/main/resources/swagger/index.html.template]: I think `startsWith` is not supported in older browsers. Probably better to use
`apiTarget.lastIndexOf(swaggerApi.System.basePath, 0) === 0`

**Comment by kroepke** [comment]
I need to test a couple of more scenarios tomorrow, but overall this looks good!

File: graylog2-server/src/main/resources/swagger/index.html.template (hunk 2 above)
Comment (kroepke) [inline graylog2-server/src/main/resources/swagger/index.html.template]: 🎯

File: graylog2-server/src/main/java/org/graylog2/shared/rest/resources/system/RemoteDeflectorResource.java (hunk 1 above)
Comment (kroepke) [inline graylog2-server/src/main/java/org/graylog2/shared/rest/resources/system/RemoteDeflectorResource.java]: We agreed to simply fix the affected resources in the relevant plugins.
Since the default is to run the API on a sub path, this will fail early on which is good enough.

**Comment by kroepke** [comment]
I've tested the following combinations with 2 nodes:
- out of the box defaults (only adjusting ports for running on the same box)
- API on `/api` (node 1) and `/schmapi` (node 2), web default
- same as above, web on `/web`
- using all current plugins

The only thing that currently does not work is to put the web listener on a separate IP/Port. The assets then all end up in a 404.

**Comment by kroepke** [comment]
Works now on the separate listener.

**Comment by kroepke** [comment]
lgtm
